### DISCUSSION TITLE
AttributeError: 'NoneType' object has no attribute 'set_grad_enabled'

### DISCUSSION THREAD

**Discussion by dai-ichiro** [awslabs/gluonts#1804] [Q&A]
When I ran the following code, figure was drawn fine but I got the error.
Please tell me how to solve it.
Is it OK to ignore it?
Thank you so much.

**To Reproduce**
```
import warnings
warnings.simplefilter('ignore')

from gluonts.dataset.common import ListDataset
from gluonts.dataset.util import to_pandas
from gluonts.torch.model.deepar import DeepAREstimator
from gluonts.evaluation.backtest import make_evaluation_predictions

import pandas as pd
from matplotlib import pyplot as plt
url = "https://raw.githubusercontent.com/numenta/NAB/master/data/realTweets/Twitter_volume_AMZN.csv"
df = pd.read_csv(url, header=0, index_col=0)

train_data = ListDataset([{
    "start": df.index[0],
    "target": df.value[:-24]
    }],
    freq="5min")

test_data = ListDataset([{
    "start": df.index[0],
    "target": df.value
    }],
    freq="5min")

estimator = DeepAREstimator(
    freq='5min',
    prediction_length=24,
    trainer_kwargs=dict(max_epochs=3),
)

predictor = estimator.train(train_data)

forecast_it, ts_it = make_evaluation_predictions(
    dataset=test_data,      
    predictor=predictor,    
    num_samples=100
) 

plot_length = 100
prediction_intervals = (50.0, 90.0)
legend = ["observations", "median prediction"] + [f"{k}% prediction interval" for k in prediction_intervals][::-1]

for x, y in zip(test_data, forecast_it):
    to_pandas(x)[-plot_length:].plot()
    y.plot(color='g', prediction_intervals=prediction_intervals)
    plt.grid(which='both')
    plt.legend(legend, loc='upper left')

plt.show()
```

**Error Message**
Exception ignored in: <generator object PyTorchPredictor.predict at 0x7f811663f660>
Traceback (most recent call last):
  File "/home/username/gluonts/lib/python3.8/site-packages/gluonts/torch/model/predictor.py", line 82, in predict
  File "/home/username/gluonts/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 129, in __exit__
AttributeError: 'NoneType' object has no attribute 'set_grad_enabled'

**Environment**
OS: Ubuntu 20.04 on WSL2
Python version: 3.8.10
gluonts==0.8.1
torch==1.10.0+cu113
pytorch-lightning==1.5.3

**Comment by MattiasDC** [comment]
The exception occurs while the Python process is shutting down. The generator `forecast_it` is not yet fully consumed and is destructed at shutdown (causing the exit block of `torch.no_grad` to execute), but after the module of `torch` is already partly cleaned up (`set_grad_enabled` is no longer there). If you would `del forecast_it` before/after `plt.show()`, this exception would not happen anymore.

As to why the generator is cleaned up that late (after the torch module is no longer valid), you'd have to investigate the garbage collector/module cleanup of Python itself
